From 0f0aa0a4abded3a5360361ec02119183710ced35 Mon Sep 17 00:00:00 2001
From: Ashutosh Chauhan <hashutosh@apache.org>
Date: Fri, 15 Jun 2012 21:23:33 +0000
Subject: [PATCH 092/148] HIVE-2796 [jira] Support auto completion for hive configs in CliDriver
 (Navis Ryu via Ashutosh Chauhan)

Summary:
DPAL-747 Support auto completion for hive configs in CliDriver

It's very cumbersome to memorize hive conf vars.

Test Plan: EMPTY

Reviewers: JIRA, ashutoshc

Reviewed By: ashutoshc

Differential Revision: https://reviews.facebook.net/D1689

git-svn-id: https://svn.apache.org/repos/asf/hive/trunk@1350807 13f79535-47bb-0310-9956-ffa450edef68
(cherry picked from commit 95393432cdd5e722f28ff41b440147eccab31554)
---
 .../java/org/apache/hadoop/hive/cli/CliDriver.java |   33 ++++++++++++++++++--
 1 files changed, 30 insertions(+), 3 deletions(-)

diff --git a/src/cli/src/java/org/apache/hadoop/hive/cli/CliDriver.java b/src/cli/src/java/org/apache/hadoop/hive/cli/CliDriver.java
index 34fc843..ed65161 100644
--- a/src/cli/src/java/org/apache/hadoop/hive/cli/CliDriver.java
+++ b/src/cli/src/java/org/apache/hadoop/hive/cli/CliDriver.java
@@ -512,7 +512,7 @@ public class CliDriver {
     }
   }
 
-  public static Completor getCommandCompletor () {
+  public static Completor[] getCommandCompletor () {
     // SimpleCompletor matches against a pre-defined wordlist
     // We start with an empty wordlist and build it up
     SimpleCompletor sc = new SimpleCompletor(new String[0]);
@@ -575,7 +575,32 @@ public class CliDriver {
       }
     };
 
-    return completor;
+    HiveConf.ConfVars[] confs = HiveConf.ConfVars.values();
+    String[] vars = new String[confs.length];
+    for (int i = 0; i < vars.length; i++) {
+      vars[i] = confs[i].varname;
+    }
+    SimpleCompletor conf = new SimpleCompletor(vars);
+    conf.setDelimiter(".");
+
+    SimpleCompletor set = new SimpleCompletor("set") {
+      @Override
+      public int complete(String buffer, int cursor, List clist) {
+        return buffer != null && buffer.equals("set") ? super.complete(buffer, cursor, clist) : -1;
+      }
+    };
+    ArgumentCompletor propCompletor = new ArgumentCompletor(new Completor[]{set, conf}) {
+      @Override
+      @SuppressWarnings("unchecked")
+      public int complete(String buffer, int offset, List completions) {
+        int ret = super.complete(buffer, offset, completions);
+        if (completions.size() == 1) {
+          completions.set(0, ((String)completions.get(0)).trim());
+        }
+        return ret;
+      }
+    };
+    return new Completor[] {propCompletor, completor};
   }
 
   public static void main(String[] args) throws Exception {
@@ -682,7 +707,9 @@ public class CliDriver {
     ConsoleReader reader = new ConsoleReader();
     reader.setBellEnabled(false);
     // reader.setDebug(new PrintWriter(new FileWriter("writer.debug", true)));
-    reader.addCompletor(getCommandCompletor());
+    for (Completor completor : getCommandCompletor()) {
+      reader.addCompletor(completor);
+    }
 
     String line;
     final String HISTORYFILE = ".hivehistory";
-- 
1.7.0.4

